src/core.js

```diff
@@ -539,45 +509,61 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
      * Makes sure there are empty rows at the bottom of the table
      */
     adjustRowsAndCols() {
-      if (priv.settings.minRows) {
+      if (tableMeta.minRows) {
         // should I add empty rows to data source to meet minRows?
         const rows = instance.countRows();
 
-        if (rows < priv.settings.minRows) {
-          for (let r = 0, minRows = priv.settings.minRows; r < minRows - rows; r++) {
-            datamap.createRow(instance.countRows(), 1, 'auto');
+        if (rows < tableMeta.minRows) {
+          for (let r = 0, minRows = tableMeta.minRows; r < minRows - rows; r++) {
+            const visualLastRowIndex = instance.countRows();
+            const physicalLastRowIndex = instance.toPhysicalRow(visualLastRowIndex);
+
+            datamap.createRow(visualLastRowIndex, 1, 'auto');
```

Comment:
After investigation, I came to the conclusion that sync via `metaManager.createColumn/createRow`  introduced unwanted behavior and should be removed. The logic here is executed for `minRows`, `minSpareRows`, `minCols` and `minSpareCols` options. Right now when the developer defines cell meta for not existed row e.g `hot.getCellMeta(8, 0).className = 'bg-red';`  and than do update settings `updateSettings({ minSpareRows: 10 })` the oryginally defined cell will be shifted to `(18, 0)`, which points to not existed cell again. 

The previous code works better for that use case. It didn't shift the cell after creating new rows/columns for the abovementioned options.

I'll revert my change and add missing tests.

Reply:
I'm not sure if it was a better behavior. Maybe we shouldn't allow setting cellMeta for non-existing cells? Cell meta should always shift. If I'm inserting new rows, I want a clear meta for them. After all, they're brand new and clean.  Suddenly I will get a red background for one of them?

Reply:
It's true for cell meta objects which have their corresponding cell data. For those 4 options, the behavior is changing. It's more convincing for me that I define cell meta for not existed row, and as soon as the logic behind that 4 options fill that row, I got my rendered options.

I got a red background for the cell, but only when one of those options is enabled. Otherwise, the cell meta objects are shifted.

Reply:
Resolved within https://github.com/handsontable/handsontable/commit/20c9937ef47946ad68f8c00767cab340d9b4f88a